File: 1-500/0210. Course Schedule II.java
// Medium
// BFS, Graph（DAG）
// Time:O(V+E), Space:O(V+E)
// https://leetcode.cn/problems/course-schedule-ii/

import java.util.*;

class Solution {
    // Use BFS to determine a valid course order
    // Build a graph and inDegree array: 
    //   - Graph stores dependencies, inDegree tracks prerequisite counts
    // Find courses with no prerequisites and process them:
    //   - For each course, reduce inDegree of its neighbors, add to que if inDegree becomes 0
    // If all courses are processed, return the order; otherwise, return empty
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        int[] inDegree = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] pre : prerequisites) {
            int course = pre[0];
            int prereq = pre[1];
            graph.get(prereq).add(course);
            inDegree[course]++;
        }
        Queue<Integer> que = new LinkedList<>();
        // find the course with no prereq as start
        for (int i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0) {
                que.offer(i);
            }
        }
        int finish = 0;
        int[] res = new int[numCourses];
        while (!que.isEmpty()) {
            int curCourse = que.poll();
            res[finish] = curCourse;
            finish++;
            for (int nextCourse : graph.get(curCourse)) {
                inDegree[nextCourse]--;
                if (inDegree[nextCourse] == 0) {
                    que.offer(nextCourse);
                }
            }
        }
        if (finish != numCourses) {
            return new int[]{}; // return empty list
        } else {
            return res;
        }
    }
}

File: 0001-0500/0210. Course Schedule II.py
# Time:O(V+E), Space:O(V+E)

from collections import deque
from typing import List

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = [[] for _ in range(numCourses)]
        indegree = [0] * numCourses

        for a, b in prerequisites:
            graph[b].append(a)
            indegree[a] += 1

        que = deque([i for i in range(numCourses) if indegree[i] == 0])
        res = []

        while que:
            node = que.popleft()
            res.append(node)
            for neighbor in graph[node]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    que.append(neighbor)

        return res if len(res) == numCourses else []
    